packages/react-router/README.md

```diff
@@ -0,0 +1,378 @@
+
+```tsx
+// Back one page
+router.go(-1);
+
+// Forward three pages
+router.go(3);
+```
+
+#### `router.forward()`
+
+An alias for `router.go(1)`.
+
+#### `router.back()`
+
+An alias for `router.go(-1)`.
+
+#### `router.block()`
+
+Blocks the router from performing additional navigations. In general, avoid this method, and use the `useNavigationBlock` hook or `NavigationBlock` component instead.
+
+#### `router.listen()`
+
+Subscribes to changes in the current URL. In general, avoid this method, and use the `useCurrentUrl` hook (and changes to the value it provides) instead.
+
+### `useNavigationBlock()` and `<NavigationBlock />`
+
+Applications sometimes need to block the user from being able to navigate away. While it is generally better to save the state of the page and rehydrate it when the user returns, this is not always possible. `@shopify/react-router` provides a mechanism for blocking all navigation, including the browser’s native back and forward buttons.
+
+The `useNavigationBlock` hook allows you to register a function that can block navigation. This function is called with arguments:
+
+- `url`, an `EnhancedURL` object representing the target location.
+- `redo`, a function that you can store and call at a later time that will forcibly perform the navigation, if you decide to block it.
+
+This function should return a boolean. If you return `true`, the navigation will be blocked. Otherwise, the navigation will be performed normally.
+
+Once the component using the `useNavigationBlock` hook is unmounted, the block will no longer be called.
+
+The `NavigationBlock` component behaves the same way, except that the function to determine the block is passed as the `onNavigation` prop.
+
+```tsx
+import {useEffect} from 'react';
+import {useNavigationBlock} from '@shopify/react-router';
+
+function Blocker() {
+  const redo = React.useRef<(() => void) | null>(null);
+
+  useNavigationBlock((url, redoNavigation) => {
+    redo.current = redoNavigation;
+    return !url.pathname.startsWith('/can/always/go/here');
+  });
+
+  useEffect(() => {
+    return () => {
+      if (redo.current != null) {
+        redo.current();
+      }
+    };
+  }, []);
+
+  return null;
+}
+```
+
+### `<Redirect />`
+
+The `Redirect` component is essentially a component wrapper around running `router.navigate(to, {replace: true})`. It accepts a `to` prop, which can be anything you can pass to `router.navigate`. This component also works with [`@shopify/react-network`](https://github.com/Shopify/quilt/tree/master/packages/react-network) to perform a network redirect when rendered on the server.
+
+### `<NoMatch />`
+
+The `NoMatch` component keeps track of whether any nested routes have been matched and, if no matches are found, renders the result of calling `renderFallback`. Note that the `Route` components do not need to be direct children; they can be nested or composed however you like.
```

Comment:
I think this is the only place we mention nested routes, how do you invision those working?

Reply:
They can work however the developer likes, just like nested components. In theory, this is now totally OK:

```tsx
<NoMatch renderFallback={() => <NotFound />}>
  <Route match="/" render={() => <Home />
  <ProductRoutes />
</NoMatch>

function ProductRoutes() {
  return <>
    <Route match="/products" render={() => <ProductList />} />
    <Route match="/products/new" render={() => <ProductDetails />} />
  </>;
}
```

So there really isn't anything new to learn, and you don't need to worry about keeping the routes on a single level like with react-router.